packages/fxa-content-server/app/scripts/models/account.js

```diff
@@ -372,6 +382,128 @@ const Account = Backbone.Model.extend(
+     *   @param {String} [options.password] Optional The user's password
+     * @returns {Promise}
+     */
+    async generateEcosystemAnonIdIfRequired(options = {}) {
+      try {
+        let ecosystemAnonId = this.get('ecosystemAnonId');
+        if (ecosystemAnonId) {
+          return;
+        }
+
+        // If no data pipeline keys are available, we can't generate the ecosystemAnonId, abort
+        if (this._ecosystemAnonIdPublicKeys.length <= 0) {
+          return;
+        }
+
+        // Pick a random public key
+        const randomEcosystemAnonIdPublicKey = this._ecosystemAnonIdPublicKeys[
+          Math.floor(Math.random() * this._ecosystemAnonIdPublicKeys.length)
+        ];
+
+        //
+        if (options.kB) {
+          ecosystemAnonId = await ecosystemAnonKeys.generateEcosystemAnonID(
+            options.kB,
+            randomEcosystemAnonIdPublicKey
+          );
+        } else if (this.canFetchKeys()) {
+          const keys = await this.accountKeys();
+          ecosystemAnonId = await ecosystemAnonKeys.generateEcosystemAnonID(
+            keys.kB,
+            randomEcosystemAnonIdPublicKey
+          );
+        } else if (options.password) {
+          // TODO: Using sessionReauth to get keys could potentially pose some issues if the user
+          // is attempting to come from an unverified session. FxA only allows fetching keys from
+          // sessions that are verified (ex. performed some email verification loop).
+          const res = await this._fxaClient.sessionReauth(
+            this.get('sessionToken'),
+            this.get('email'),
+            options.password,
+            {
+              wantsKeys: () => true, // Ensure keys are requested
+              has: () => false,
+              isSync: () => false,
+            },
+            {
+              keys: true,
+
+              // TODO: The verification reason needs to be updated to be specific for ecosystemAnonId
+              // There are checks in the auth-server to send emails depending on the `reason` value specified.
+              reason: VerificationReasons.RECOVERY_KEY,
+            }
+          );
+
+          if (res.keyFetchToken && res.unwrapBKey) {
+            const keys = await this._fxaClient.accountKeys(
+              res.keyFetchToken,
+              res.unwrapBKey
+            );
+            ecosystemAnonId = await ecosystemAnonKeys.generateEcosystemAnonID(
+              keys.kB,
+              randomEcosystemAnonIdPublicKey
+            );
+          }
+        }
+
+        // Update localstorage and report new ecosystemAnonId to auth-server
+        if (ecosystemAnonId) {
+          this.set('ecosystemAnonId', ecosystemAnonId);
+
+          await this._fxaClient.updateEcosystemAnonId(
+            this.get('sessionToken'),
+
+            // TODO: We need to ensure that the ecosystemAnonId is stored and updated in correct format.
+            // ie. hex or base64
```

Comment:
Hex format is only needed for submission to the data pipeline. Everywhere else we can stick with the serialization suggested in the ['Anonymizing User Identifiers' AET doc](https://docs.google.com/document/d/1zH3eVVI_28Afg1JXe_McDrW4MTYuWhiJMQR6AQbli8I/edit), which is the [JWE compact serialization](https://tools.ietf.org/html/rfc7516#section-7.1) in url-safe base64. I'll add a note to the new bug around submitting to the data pipeline, and I think we're good to delete this TODO :+1: